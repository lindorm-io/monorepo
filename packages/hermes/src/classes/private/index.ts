export * from "./HermesMetadata";
export * from "./HermesRegistry";
export * from "./HermesScanner";
