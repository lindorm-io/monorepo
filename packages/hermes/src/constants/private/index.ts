export * from "./checksum-store";
export * from "./domain";
export * from "./event-store";
export * from "./saga-store";
export * from "./view-store";
