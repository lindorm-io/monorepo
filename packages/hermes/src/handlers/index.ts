export * from "./HermesAggregateCommandHandler";
export * from "./HermesAggregateErrorHandler";
export * from "./HermesAggregateEventHandler";
export * from "./HermesSagaErrorHandler";
export * from "./HermesSagaEventHandler";
export * from "./HermesSagaIdHandler";
export * from "./HermesSagaTimeoutHandler";
export * from "./HermesViewErrorHandler";
export * from "./HermesViewEventHandler";
export * from "./HermesViewIdHandler";
export * from "./HermesViewQueryHandler";
