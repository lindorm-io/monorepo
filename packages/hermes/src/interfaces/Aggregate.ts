import { ClassLike, Dict } from "@lindorm/types";
import { AggregateData } from "../types";
import { IHermesMessage } from "./HermesMessage";

export interface IAggregate<S extends Dict = Dict> extends AggregateData {
  apply(causation: IHermesMessage, event: ClassLike): Promise<void>;
  load(event: IHermesMessage): void;
  toJSON(): AggregateData<S>;
}
