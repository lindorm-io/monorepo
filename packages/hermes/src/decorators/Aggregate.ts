import { Constructor } from "@lindorm/types";
import { AggregateDecoratorOptions } from "../types";
import { extractNameData, globalHermesMetadata } from "../utils/private";

export function Aggregate(options: AggregateDecoratorOptions = {}): ClassDecorator {
  return function (target) {
    const { name } = extractNameData(target.name);
    globalHermesMetadata.addAggregate({
      encryption: options.encryption || false,
      name: options.name || name,
      namespace: options.namespace || null,
      target: target as unknown as Constructor,
    });
  };
}
