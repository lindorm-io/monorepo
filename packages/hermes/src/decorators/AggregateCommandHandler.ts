import { Constructor, Dict } from "@lindorm/types";
import {
  AggregateCommandHandlerDecoratorOptions,
  AggregateCommandHandlerDescriptor,
} from "../types";
import { globalHermesMetadata } from "../utils/private";

export function AggregateCommandHandler<C extends Constructor, S extends Dict>(
  CommandClass: C,
  options: AggregateCommandHandlerDecoratorOptions = {},
): AggregateCommandHandlerDescriptor<C, S> {
  return function (target, key, descriptor) {
    globalHermesMetadata.addHandler({
      conditions: options.conditions || null,
      decorator: "AggregateCommandHandler",
      encryption: options.encryption || false,
      handler: descriptor.value,
      key: key.toString(),
      schema: options.schema || null,
      target: target.constructor as Constructor,
      trigger: CommandClass,
    });
  };
}
