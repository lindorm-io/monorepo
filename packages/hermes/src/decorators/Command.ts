import { Constructor } from "@lindorm/types";
import { CommandDecoratorOptions } from "../types";
import { extractNameData, globalHermesMetadata } from "../utils/private";

export function Command(options?: CommandDecoratorOptions): ClassDecorator {
  return function (target) {
    const { name } = extractNameData(target.name);
    globalHermesMetadata.addCommand({
      name: options?.name || name,
      target: target as unknown as Constructor,
    });
  };
}
