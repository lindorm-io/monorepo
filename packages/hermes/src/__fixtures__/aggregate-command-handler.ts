import { z } from "zod";
import { HermesAggregateCommandHandler } from "../handlers";
import { AggregateCommandHandlerOptions } from "../types";
import { TEST_AGGREGATE_OPTIONS } from "./aggregate";
import {
  TEST_HERMES_COMMAND,
  TEST_HERMES_COMMAND_ADD_FIELD,
  TEST_HERMES_COMMAND_CREATE,
  TEST_HERMES_COMMAND_DESTROY,
  TEST_HERMES_COMMAND_DESTROY_NEXT,
  TEST_HERMES_COMMAND_DISPATCH,
  TEST_HERMES_COMMAND_ENCRYPT,
  TEST_HERMES_COMMAND_MERGE_STATE,
  TEST_HERMES_COMMAND_REMOVE_FIELD_WHERE_EQUAL,
  TEST_HERMES_COMMAND_REMOVE_FIELD_WHERE_MATCH,
  TEST_HERMES_COMMAND_SET_STATE,
  TEST_HERMES_COMMAND_THROWS,
  TEST_HERMES_COMMAND_TIMEOUT,
} from "./hermes-command";

export const TEST_AGGREGATE_COMMAND_HANDLER_OPTIONS: AggregateCommandHandlerOptions = {
  aggregate: {
    name: TEST_AGGREGATE_OPTIONS.name,
    context: TEST_AGGREGATE_OPTIONS.context,
  },
  commandName: TEST_HERMES_COMMAND.name,
  conditions: { created: true },
  schema: z.object({
    commandData: z.boolean(),
  }),
  handler: jest.fn().mockImplementation(async () => {}),
};

export const TEST_AGGREGATE_COMMAND_HANDLER = new HermesAggregateCommandHandler(
  TEST_AGGREGATE_COMMAND_HANDLER_OPTIONS,
);

export const TEST_AGGREGATE_COMMAND_HANDLER_CREATE = new HermesAggregateCommandHandler({
  ...TEST_AGGREGATE_COMMAND_HANDLER_OPTIONS,
  commandName: TEST_HERMES_COMMAND_CREATE.name,
  conditions: { created: false },
  handler: jest.fn().mockImplementation(async (ctx) => {
    class HermesEventCreate {
      constructor(public readonly dataFromCommand: any) {}
    }
    await ctx.apply(new HermesEventCreate(ctx.command));
  }),
});

export const TEST_AGGREGATE_COMMAND_HANDLER_ADD_FIELD = new HermesAggregateCommandHandler(
  {
    ...TEST_AGGREGATE_COMMAND_HANDLER_OPTIONS,
    commandName: TEST_HERMES_COMMAND_ADD_FIELD.name,
    handler: jest.fn().mockImplementation(async (ctx) => {
      class HermesEventAddField {
        constructor(public readonly dataFromCommand: any) {}
      }
      await ctx.apply(new HermesEventAddField(ctx.command));
    }),
  },
);

export const TEST_AGGREGATE_COMMAND_HANDLER_DESTROY = new HermesAggregateCommandHandler({
  ...TEST_AGGREGATE_COMMAND_HANDLER_OPTIONS,
  commandName: TEST_HERMES_COMMAND_DESTROY.name,
  handler: jest.fn().mockImplementation(async (ctx) => {
    class HermesEventDestroy {
      constructor(public readonly dataFromCommand: any) {}
    }
    await ctx.apply(new HermesEventDestroy(ctx.command));
  }),
});

export const TEST_AGGREGATE_COMMAND_HANDLER_DESTROY_NEXT =
  new HermesAggregateCommandHandler({
    ...TEST_AGGREGATE_COMMAND_HANDLER_OPTIONS,
    commandName: TEST_HERMES_COMMAND_DESTROY_NEXT.name,
    handler: jest.fn().mockImplementation(async (ctx) => {
      class HermesEventDestroyNext {
        constructor(public readonly dataFromCommand: any) {}
      }
      await ctx.apply(new HermesEventDestroyNext(ctx.command));
    }),
  });

export const TEST_AGGREGATE_COMMAND_HANDLER_DISPATCH = new HermesAggregateCommandHandler({
  ...TEST_AGGREGATE_COMMAND_HANDLER_OPTIONS,
  commandName: TEST_HERMES_COMMAND_DISPATCH.name,
  handler: jest.fn().mockImplementation(async (ctx) => {
    class HermesEventDispatch {
      constructor(public readonly dataFromCommand: any) {}
    }
    await ctx.apply(new HermesEventDispatch(ctx.command));
  }),
});

export const TEST_AGGREGATE_COMMAND_HANDLER_ENCRYPT = new HermesAggregateCommandHandler({
  ...TEST_AGGREGATE_COMMAND_HANDLER_OPTIONS,
  commandName: TEST_HERMES_COMMAND_ENCRYPT.name,
  conditions: {},
  encryption: true,
  handler: jest.fn().mockImplementation(async (ctx) => {
    class HermesEventEncrypt {
      constructor(public readonly encryptedData: any) {}
    }
    await ctx.apply(new HermesEventEncrypt(ctx.command));
  }),
});

export const TEST_AGGREGATE_COMMAND_HANDLER_MERGE_STATE =
  new HermesAggregateCommandHandler({
    ...TEST_AGGREGATE_COMMAND_HANDLER_OPTIONS,
    commandName: TEST_HERMES_COMMAND_MERGE_STATE.name,
    handler: jest.fn().mockImplementation(async (ctx) => {
      class HermesEventMergeState {
        constructor(public readonly dataFromCommand: any) {}
      }
      await ctx.apply(new HermesEventMergeState(ctx.command));
    }),
  });

export const TEST_AGGREGATE_COMMAND_HANDLER_REMOVE_FIELD_WHERE_EQUAL =
  new HermesAggregateCommandHandler({
    ...TEST_AGGREGATE_COMMAND_HANDLER_OPTIONS,
    commandName: TEST_HERMES_COMMAND_REMOVE_FIELD_WHERE_EQUAL.name,
    handler: jest.fn().mockImplementation(async (ctx) => {
      class HermesEventRemoveFieldWhereEqual {
        constructor(public readonly dataFromCommand: any) {}
      }
      await ctx.apply(new HermesEventRemoveFieldWhereEqual(ctx.command));
    }),
  });

export const TEST_AGGREGATE_COMMAND_HANDLER_REMOVE_FIELD_WHERE_MATCH =
  new HermesAggregateCommandHandler({
    ...TEST_AGGREGATE_COMMAND_HANDLER_OPTIONS,
    commandName: TEST_HERMES_COMMAND_REMOVE_FIELD_WHERE_MATCH.name,
    handler: jest.fn().mockImplementation(async (ctx) => {
      class HermesEventRemoveFieldWhereMatch {
        constructor(public readonly dataFromCommand: any) {}
      }
      await ctx.apply(new HermesEventRemoveFieldWhereMatch(ctx.command));
    }),
  });

export const TEST_AGGREGATE_COMMAND_HANDLER_SET_STATE = new HermesAggregateCommandHandler(
  {
    ...TEST_AGGREGATE_COMMAND_HANDLER_OPTIONS,
    commandName: TEST_HERMES_COMMAND_SET_STATE.name,
    handler: jest.fn().mockImplementation(async (ctx) => {
      class HermesEventSetState {
        constructor(public readonly dataFromCommand: any) {}
      }
      await ctx.apply(new HermesEventSetState(ctx.command));
    }),
  },
);

export const TEST_AGGREGATE_COMMAND_HANDLER_THROWS = new HermesAggregateCommandHandler({
  ...TEST_AGGREGATE_COMMAND_HANDLER_OPTIONS,
  commandName: TEST_HERMES_COMMAND_THROWS.name,
  handler: jest.fn().mockImplementation(async (ctx) => {
    class HermesEventThrows {
      constructor(public readonly dataFromCommand: any) {}
    }
    await ctx.apply(new HermesEventThrows(ctx.command));
  }),
});

export const TEST_AGGREGATE_COMMAND_HANDLER_TIMEOUT = new HermesAggregateCommandHandler({
  ...TEST_AGGREGATE_COMMAND_HANDLER_OPTIONS,
  commandName: TEST_HERMES_COMMAND_TIMEOUT.name,
  handler: jest.fn().mockImplementation(async (ctx) => {
    class HermesEventTimeout {
      constructor(public readonly dataFromCommand: any) {}
    }
    await ctx.apply(new HermesEventTimeout(ctx.command));
  }),
});
