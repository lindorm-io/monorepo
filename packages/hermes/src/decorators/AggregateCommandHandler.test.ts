import { Dict } from "@lindorm/types";
import z from "zod";
import { AggregateCommandCtx } from "../types/handlers/aggregate-command-handler";
import { globalHermesMetadata } from "../utils/private";
import { Aggregate } from "./Aggregate";
import { AggregateCommandHandler } from "./AggregateCommandHandler";
import { Command } from "./Command";

describe("AggregateCommandHandler Decorator", () => {
  test("should add metadata", () => {
    @Command()
    class TestCommand {}

    @Aggregate()
    class TestAggregateCommandHandlerAggregate {
      @AggregateCommandHandler(TestCommand)
      public async onTestAggregateDomainCommand(
        ctx: AggregateCommandCtx<TestCommand, Dict>,
      ) {}
    }

    expect(
      globalHermesMetadata.getAggregate(TestAggregateCommandHandlerAggregate),
    ).toMatchSnapshot();
  });

  test("should add metadata with custom options", () => {
    @Command()
    class TestCommand {}

    @Aggregate()
    class TestAggregateCommandHandlerAggregate {
      @AggregateCommandHandler(TestCommand, {
        conditions: { created: true },
        encryption: true,
        schema: z.object({}),
      })
      public async onTestAggregateDomainCommand(
        ctx: AggregateCommandCtx<TestCommand, Dict>,
      ) {}
    }

    expect(
      globalHermesMetadata.getAggregate(TestAggregateCommandHandlerAggregate),
    ).toMatchSnapshot();
  });
});
