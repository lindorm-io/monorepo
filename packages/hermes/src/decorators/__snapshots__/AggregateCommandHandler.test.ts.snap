// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AggregateCommandHandler Decorator should add metadata 1`] = `
{
  "encryption": false,
  "handlers": [
    {
      "conditions": null,
      "decorator": "AggregateCommandHandler",
      "encryption": false,
      "handler": [Function],
      "key": "onTestAggregateDomainCommand",
      "schema": null,
      "target": [Function],
      "trigger": [Function],
    },
  ],
  "name": "test_aggregate_command_handler_aggregate",
  "namespace": null,
  "target": [Function],
}
`;

exports[`AggregateCommandHandler Decorator should add metadata with custom options 1`] = `
{
  "encryption": false,
  "handlers": [
    {
      "conditions": {
        "created": true,
      },
      "decorator": "AggregateCommandHandler",
      "encryption": true,
      "handler": [Function],
      "key": "onTestAggregateDomainCommand",
      "schema": ZodObject {
        "_cached": null,
        "_def": {
          "catchall": ZodNever {
            "_def": {
              "typeName": "ZodNever",
            },
            "and": [Function],
            "array": [Function],
            "brand": [Function],
            "catch": [Function],
            "default": [Function],
            "describe": [Function],
            "isNullable": [Function],
            "isOptional": [Function],
            "nullable": [Function],
            "nullish": [Function],
            "optional": [Function],
            "or": [Function],
            "parse": [Function],
            "parseAsync": [Function],
            "pipe": [Function],
            "promise": [Function],
            "readonly": [Function],
            "refine": [Function],
            "refinement": [Function],
            "safeParse": [Function],
            "safeParseAsync": [Function],
            "spa": [Function],
            "superRefine": [Function],
            "transform": [Function],
            "~standard": {
              "validate": [Function],
              "vendor": "zod",
              "version": 1,
            },
          },
          "shape": [Function],
          "typeName": "ZodObject",
          "unknownKeys": "strip",
        },
        "and": [Function],
        "array": [Function],
        "augment": [Function],
        "brand": [Function],
        "catch": [Function],
        "default": [Function],
        "describe": [Function],
        "isNullable": [Function],
        "isOptional": [Function],
        "nonstrict": [Function],
        "nullable": [Function],
        "nullish": [Function],
        "optional": [Function],
        "or": [Function],
        "parse": [Function],
        "parseAsync": [Function],
        "pipe": [Function],
        "promise": [Function],
        "readonly": [Function],
        "refine": [Function],
        "refinement": [Function],
        "safeParse": [Function],
        "safeParseAsync": [Function],
        "spa": [Function],
        "superRefine": [Function],
        "transform": [Function],
        "~standard": {
          "validate": [Function],
          "vendor": "zod",
          "version": 1,
        },
      },
      "target": [Function],
      "trigger": [Function],
    },
  ],
  "name": "test_aggregate_command_handler_aggregate",
  "namespace": null,
  "target": [Function],
}
`;
